src/app/shared/daterange/daterange.component.ts

```diff
@@ -0,0 +1,226 @@
+    </mat-form-field>
+    <button mat-raised-button (click)="openPicker(fromPicker)">Open</button>
+    <grove-daterange-overlay [display]="showBackdrop"></grove-daterange-overlay>
+  `,
+  providers: [
+    {
+      provide: NG_VALUE_ACCESSOR,
+      useExisting: forwardRef(() => DaterangeComponent),
+      multi: true
+    }
+  ],
+  styleUrls: ['daterange.component.scss'],
+  encapsulation: ViewEncapsulation.None
+})
+export class DaterangeComponent implements AfterViewInit, OnDestroy, ControlValueAccessor {
+  @ViewChild('fromPicker') fromPicker: MatDatepicker<Date>;
+  @ViewChild('toPicker') toPicker: MatDatepicker<Date>;
+  @ViewChild(MatDatepickerInput) fromInput: MatDatepickerInput<Date>;
+  @Input() _startDate = null;
+  @Input() _endDate = null;
+
+  set startDate(val: Date) {
+    this._startDate = val;
+    this.propagateChange({startDate: this.startDate, endDate: this.endDate});
+  }
+
+  get startDate() {
+    return this._startDate;
+  }
+
+  set endDate(val: Date) {
+    this._endDate = val;
+    this.propagateChange({startDate: this.startDate, endDate: this.endDate});
+  }
+
+  get endDate() {
+    return this._endDate;
+  }
+
+  authToken: string;
+  rafRegistry = {};
+  overlay: Overlay;
+  overlays: OverlayRef[] = [];
+  showBackdrop = false;
+  get toPickerInitialDate() {
+    if (this.startDate) { return this.startDate; }
+    return new Date();
+  }
+
+  propagateChange = (_: any) => {};
+
+  constructor(overlay: Overlay, private _injector: Injector) {
+    this.overlay = overlay;
+  }
+
+  writeValue(val: {startDate?: Date, endDate?: Date} | null) {
+    if (val && val.startDate) { this.startDate = val.startDate; }
+    if (val && val.endDate) { this.endDate = val.endDate; }
+  }
+
+  registerOnChange(fn) {
+    this.propagateChange = fn;
+  }
+
+  registerOnTouched() {
+
+  }
+
+  startDateChangeEvent(type: string, event: MatDatepickerInputEvent<Date>) {
+    this.startDate = event.value;
+  }
+  endDateChangeEvent(type: string, event: MatDatepickerInputEvent<Date>) {
+    this.endDate = event.value;
+    this.overlays.forEach(overlay => overlay.dispose());
+    this.showBackdrop = false;
+  }
+  // Ultimately, we'd like to keep both pickers open until both dates are
+  // selected. This is an open feature request for Angular Material:
+  // https://github.com/angular/components/issues/15295
+  openPicker(fromPicker: MatDatepicker<Date>) {
+    this.startDate = null;
+    fromPicker.select(null);
+    this.toPicker.select(null);
+    fromPicker.open();
+  }
+
+  highlightStartDate = (d: Date) => {
+    if (!this.startDate) {
+      return undefined;
+    }
+
+    return this.checkDateEquality([d, this.startDate]) ? 'start-date' : undefined;
+  };
+
+  rangeFilter = (d: Date): boolean => {
+    return this.startDate ? this.startDate <= d : false;
+  };
+
+  checkDateEquality = (dates: Date[]) => {
```

Comment:
how clever. sorry but I'm gonna change this to be more readable and actually type safe and just call the getDate, getMonth, getFullYear functions and check equality.